---
title: "Testing"
author:
  - name: "Emir Turkes and Naoto Watamura, UK DRI at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: show
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile,
    encoding = encoding,
    output_file = file.path("..", "results", "testing.html")
  )})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

```{r}
# Copyright 2025 Emir Turkes, Naoto Watamura, UK DRI at UCL
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Load packages.
# --------------
library(conflicted)
conflicts_prefer(edgeR::cpm, .quiet = TRUE)
packages <- c(
  "Seurat", "GSEABase", "GSVA", "BiocParallel", "parallelly", "Rcpp", "scuttle",
  "edgeR"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# --------------

# Define functions.
# -----------------
source("utils.R")
sourceCpp(file.path("..", "src", "calculateScores.cpp"))

`%notin%` <- Negate(`%in%`)
# -----------------

# Add paths.
# ----------
data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# ----------

# Global settings.
# ----------------
knitr::opts_chunk$set(fig.width = 7, fig.height = 5, dpi = 96)
# ----------------
```

# Prep

```{r}
seurat <- readRDS(file.path(cache_dir, "snrnaseq_integrated.rds"))
DefaultAssay(seurat) <- "RNA"

rds <- file.path(cache_dir, "gene_sets.rds")
if (file.exists(rds)) {
  gene_sets <- readRDS(rds)
} else {

  gene_sets <- getGmt(
    file.path(
      data_dir, "gprofiler_mmusculus_ENSG_2025_10_22",
      "mmusculus.GO.comb.ENSG.gmt"
    )
  )

  remove <- unique(unlist(geneIds(gene_sets)))[
    unique(unlist(geneIds(gene_sets))) %notin% rownames(seurat)
  ]
  remove <- which(
    sapply(geneIds(gene_sets), FUN = function(x) any(x %in% remove))
  )
  gene_sets <- gene_sets[-remove]

  for (i in seq_along(gene_sets@.Data)) {
    go_id <- gene_sets[[i]]@setName
    suppressWarnings(gene_sets[[i]]@setName <- gene_sets[[i]]@shortDescription)
    suppressWarnings(gene_sets[[i]]@shortDescription <- go_id)
  }

  gene_sets_keep <- computeGeneSetsOverlapMax(
    gene_sets, uniqGenes = unique(unlist(geneIds(gene_sets)))
  )
  gene_sets_keep <- rowSums(gene_sets_keep)
  gene_sets_keep <- gene_sets_keep[order(gene_sets_keep, decreasing = TRUE)]
  gene_sets <- gene_sets[match(names(gene_sets_keep), names(gene_sets))]

  saveRDS(gene_sets, file = rds)
}

gene_sets

rds <- file.path(cache_dir, "gene_sets_keep.rds")
if (file.exists(rds)) {
  gene_sets_keep <- readRDS(rds)
} else {

  gene_sets_keep <- gene_sets

  dup <- sapply(
    geneIds(gene_sets_keep), FUN = function(x) paste(sort(x), collapse = "-")
  )
  dup <- which(duplicated(dup))
  gene_sets_keep <- gene_sets_keep[-dup]

  keep <- filterGeneSets(gene_sets_keep, min.sz = 2, max.sz = 30)
  gene_sets_keep <- gene_sets_keep[names(gene_sets_keep) %in% names(keep)]

  gene_sets_keep <- gene_sets_keep[
    !grepl("regulation", x = names(gene_sets_keep)) |
      grepl("positive regulation", x = names(gene_sets_keep))
  ]

  gene_sets_keep <- gene_sets_keep[
    !grepl("selection", x = names(gene_sets_keep)) |
      grepl("positive selection", x = names(gene_sets_keep))
  ]

  gene_sets_ordered <- computeGeneSetsOverlapMax(
    gene_sets_keep, uniqGenes = unique(unlist(geneIds(gene_sets_keep)))
  )
  gene_sets_ordered <- rowSums(gene_sets_ordered)
  gene_sets_ordered <- gene_sets_ordered[order(gene_sets_ordered)]
  gene_sets_keep <- gene_sets_keep[
    match(names(gene_sets_ordered), names(gene_sets_keep))
  ]

  gene_sets_keep <- computeGeneSetsOverlapMax(
    gene_sets_keep, uniqGenes = unique(unlist(geneIds(gene_sets_keep)))
  )
  gene_sets_keep[upper.tri(gene_sets_keep)] <- 0
  diag(gene_sets_keep) <- 0
  gene_sets_keep <- apply(gene_sets_keep, MARGIN = 1, FUN = max)

  rm(gene_sets_ordered)
  saveRDS(gene_sets_keep, file = rds)
}

gene_sets_sub <- gene_sets[names(gene_sets) %in% names(gene_sets_keep)]
gene_sets_sub

gene_anno <- data.frame(
  symbol = unlist(seurat@misc), ensembl = rownames(seurat)
)
rds <- file.path(cache_dir, "gene_anno.rds")
if (!file.exists(rds)) {
  saveRDS(gene_anno, file = rds)
}
```

# Gene Set Scoring

```{r}
rds <- file.path(cache_dir, "final_seurat.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)
} else {

  scores <- genefunnel(
    GetAssayData(seurat, layer = "counts"), geneIds(gene_sets),
    BPPARAM = MulticoreParam(availableCores())
  )

  seurat[["GF"]] <- CreateAssay5Object(scores, data = log1p(scores))
  DefaultAssay(seurat) <- "GF"
  seurat$nCount_GF <- colSums(seurat, layer = "counts")
  seurat$nFeature_GF <- colSums(
    GetAssayData(seurat, layer = "counts") > 0
  )

  seurat$genotype_status_cleaned <- factor(
    seurat$genotype_status_cleaned,
    levels = c(
      "MAPTKI",
      "P301S",
      "NLGF_MAPTKI_Homeostatic",
      "NLGF_MAPTKI_Activated",
      "NLGF_P301S_Homeostatic",
      "NLGF_P301S_Activated"
    )
  )
  seurat$genotype_status_cleaned_batch <- factor(
    paste(seurat$genotype_status_cleaned, seurat$batch, sep = "_"),
    levels = paste(
      rep(levels(seurat$genotype_status_cleaned), each = 4),
      paste0("batch0", seq(4)), sep = "_"
    )
  )
  seurat@active.ident <- seurat$genotype_status_cleaned_batch

  saveRDS(seurat, file = rds)
}
```

# Testing

```{r}
gse <- as.SingleCellExperiment(seurat, assay = "GF")
gse <- gse[rownames(gse) %in% names(gene_sets_keep), ]
gse <- suppressWarnings(
  aggregateAcrossCells(
    gse, colData(gse)[ , "genotype_status_cleaned_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
gse <- gse[ , order(gse$genotype_status_cleaned_batch)]

deg <- as.SingleCellExperiment(seurat, assay = "RNA")
deg <- suppressWarnings(
  aggregateAcrossCells(
    deg, colData(deg)[ , "genotype_status_cleaned_batch"],
    use_exprs_values = "counts", statistics = "sum"
  )
)
deg <- deg[ , order(deg$genotype_status_cleaned_batch)]

dge_orig <- DGEList(counts(gse))
keep <- filterByExpr(
  dge_orig, group = gse$genotype_status_cleaned,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(gse))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = gse$genotype_status_cleaned,
  min.count = 1, min.total.count = 1.5
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

gse <- gse[rownames(gse) %in% rownames(dge)]
logcounts(gse, withDimnames = FALSE) <- cpm(
  dge, log = TRUE, prior.count = 0.25
)

design <- model.matrix(~ 0 + gse$genotype_status_cleaned + gse$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  P301S_MAPTKI =
    gse.genotype_status_cleanedP301S -
    gse.genotype_status_cleanedMAPTKI,
  NLGF_MAPTKI_HOMEOSTATIC_MAPTKI =
    gse.genotype_status_cleanedNLGF_MAPTKI_Homeostatic -
    gse.genotype_status_cleanedMAPTKI,
  NLGF_MAPTKI_ACTIVATED_MAPTKI =
    gse.genotype_status_cleanedNLGF_MAPTKI_Activated -
    gse.genotype_status_cleanedMAPTKI,
  NLGF_MAPTKI_ACTIVATED_NLGF_MAPTKI_HOMEOSTATIC =
    gse.genotype_status_cleanedNLGF_MAPTKI_Activated -
    gse.genotype_status_cleanedNLGF_MAPTKI_Homeostatic,
  NLGF_P301S_HOMEOSTATIC_P301S =
    gse.genotype_status_cleanedNLGF_P301S_Homeostatic -
    gse.genotype_status_cleanedP301S,
  NLGF_P301S_ACTIVATED_P301S =
    gse.genotype_status_cleanedNLGF_P301S_Activated -
    gse.genotype_status_cleanedP301S,
  NLGF_P301S_ACTIVATED_NLGF_P301S_HOMEOSTATIC =
    gse.genotype_status_cleanedNLGF_P301S_Activated -
    gse.genotype_status_cleanedNLGF_P301S_Homeostatic,
  NLGF_P301S_HOMEOSTATIC_NLGF_MAPTKI_HOMEOSTATIC =
    gse.genotype_status_cleanedNLGF_P301S_Homeostatic -
    gse.genotype_status_cleanedNLGF_MAPTKI_Homeostatic,
  NLGF_P301S_ACTIVATED_NLGF_MAPTKI_ACTIVATED =
    gse.genotype_status_cleanedNLGF_P301S_Activated -
    gse.genotype_status_cleanedNLGF_MAPTKI_Activated,
  levels = design
)

fit <- lmFit(logcounts(gse), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)
print(plotSA(fit))

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(fit, tests, file = tc, adjust = "BH")
close(tc)
results_gse <- read.delim(text = results)

rownames(results_gse) <- results_gse$X
results_gse <- results_gse[ , 2:40]
results_gse$GeneSet <- rownames(results_gse)
results_gse <- results_gse[order(results_gse$F.p.value), ]

gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(gse)]
gene_sets_sub <- unique(unlist(geneIds(gene_sets_sub)))
deg <- deg[rownames(deg) %in% gene_sets_sub, ]
dge_orig <- DGEList(counts(deg))
keep <- filterByExpr(
  dge_orig, group = deg$tau,
  min.count = 10, min.total.count = 15
)
dge_orig <- dge_orig[keep, ]
dge_orig <- calcNormFactors(dge_orig)
dge <- DGEList(counts(deg))
dge$samples$norm.factors <- dge_orig$samples$norm.factors
keep <- filterByExpr(
  dge, group = deg$tau,
  min.count = 4, min.total.count = 6
)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)

deg <- deg[rownames(deg) %in% rownames(dge)]
logcounts(deg, withDimnames = FALSE) <- cpm(
  dge, log = TRUE, prior.count = 1
)

design <- model.matrix(~ 0 + deg$genotype_status_cleaned + deg$batch)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  P301S_MAPTKI =
    deg.genotype_status_cleanedP301S -
    deg.genotype_status_cleanedMAPTKI,
  NLGF_MAPTKI_HOMEOSTATIC_MAPTKI =
    deg.genotype_status_cleanedNLGF_MAPTKI_Homeostatic -
    deg.genotype_status_cleanedMAPTKI,
  NLGF_MAPTKI_ACTIVATED_MAPTKI =
    deg.genotype_status_cleanedNLGF_MAPTKI_Activated -
    deg.genotype_status_cleanedMAPTKI,
  NLGF_MAPTKI_ACTIVATED_NLGF_MAPTKI_HOMEOSTATIC =
    deg.genotype_status_cleanedNLGF_MAPTKI_Activated -
    deg.genotype_status_cleanedNLGF_MAPTKI_Homeostatic,
  NLGF_P301S_HOMEOSTATIC_P301S =
    deg.genotype_status_cleanedNLGF_P301S_Homeostatic -
    deg.genotype_status_cleanedP301S,
  NLGF_P301S_ACTIVATED_P301S =
    deg.genotype_status_cleanedNLGF_P301S_Activated -
    deg.genotype_status_cleanedP301S,
  NLGF_P301S_ACTIVATED_NLGF_P301S_HOMEOSTATIC =
    deg.genotype_status_cleanedNLGF_P301S_Activated -
    deg.genotype_status_cleanedNLGF_P301S_Homeostatic,
  NLGF_P301S_HOMEOSTATIC_NLGF_MAPTKI_HOMEOSTATIC =
    deg.genotype_status_cleanedNLGF_P301S_Homeostatic -
    deg.genotype_status_cleanedNLGF_MAPTKI_Homeostatic,
  NLGF_P301S_ACTIVATED_NLGF_MAPTKI_ACTIVATED =
    deg.genotype_status_cleanedNLGF_P301S_Activated -
    deg.genotype_status_cleanedNLGF_MAPTKI_Activated,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)
print(plotSA(fit))

tests <- decideTests(fit, method = "global")
tc <- textConnection("results", open = "w")
write.fit(fit, tests, file = tc, adjust = "BH")
close(tc)
results_deg <- read.delim(text = results)

rownames(results_deg) <- results_deg$X
results_deg <- results_deg[ , 2:40]
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results_deg), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results_deg), gene_anno_sub$ensembl),
]
results_deg$Gene <- gene_anno_sub$symbol
results_deg <- results_deg[ , c(ncol(results_deg), 1:(ncol(results_deg) - 1))]
results_deg <- results_deg[order(results_deg$F.p.value), ]

meta <- data.frame(
  sample = gse$batch, genotype_status = gse$genotype_status_cleaned
)
gse <- removeBatchEffect(
  logcounts(gse), batch = gse$batch, group = gse$genotype_status_cleaned
)
deg <- removeBatchEffect(
  logcounts(deg), batch = deg$batch, group = deg$genotype_status_cleaned
)

rds <- file.path(cache_dir, "meta.rds")
if (!file.exists(rds)) {
  saveRDS(meta, file = rds)
}
rds <- file.path(cache_dir, "gse.rds")
if (!file.exists(rds)) {
  saveRDS(gse, file = rds)
}
rds <- file.path(cache_dir, "deg.rds")
if (!file.exists(rds)) {
  saveRDS(deg, file = rds)
}
rds <- file.path(cache_dir, "res_gse.rds")
if (!file.exists(rds)) {
  saveRDS(results_gse, file = rds)
}
rds <- file.path(cache_dir, "res_deg.rds")
if (!file.exists(rds)) {
  saveRDS(results_deg, file = rds)
}
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
